File: problems/Group Anagrams/groupAnagrams.cs
public class Solution {
    public IList<IList<string>> GroupAnagrams(string[] strs) {
        var result = new List<IList<string>>();
        var store = new Dictionary<string, int>();

        if (null == strs || 0 == strs.Length) {
            return result;
        }

        foreach (var item in strs) {
            var str = item.ToCharArray();

            Array.Sort(str);

            var key = string.Join("", str);

            if (store.ContainsKey(key)) {
                result[store[key]].Add(item);
            } else {
                result.Add(new List<string> {item});
                store.Add(key, result.Count - 1);
            }
        }

        return result;
    }
}

File: problems/Group Anagrams/groupAnagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        List<List<String>> result = new ArrayList<>();
        HashMap<String, Integer> store = new HashMap<>();

        for (String item : strs) {
            char[] letters = item.toCharArray();
            Arrays.sort(letters);
            String sortedLetters = new String(letters);

            if (store.containsKey(sortedLetters)) {
                result.get(store.get(sortedLetters)).add(item);
            } else {
                result.add(new ArrayList<>());
                store.put(sortedLetters, result.size() - 1);
                result.get(result.size() - 1).add(item);
            }
        }

        return result;
    }
}
